arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -62,19 +62,36 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) \
-	|| defined(CONFIG_ARMV7_R)
-	cpsie i
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
-	/* clear BASEPRI so wfi is awakened by incoming interrupts */
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	/*
+	 * PRIMASK is always cleared on ARMv7-M and ARMv8-M; set PRIMASK to
+	 * prevent interruption before wake-up.
+	 */
+	cpsid i
+
+	/*
+	 * Set wake-up interrupt priority to the lowest and synchronise to
+	 * ensure that this is visible for the WFI instruction execution.
+	 */
 	eors.n r0, r0
 	msr BASEPRI, r0
-#else
-#error Unknown ARM architecture
-#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+	isb
+
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+
+	/* Ensure that all memory transactions are completed */
+	dsb
 
+	/* Enter low power state */
 	wfi
 
+	/*
+	 * Clear PRIMASK and flush instruction buffer to immediately service
```

Comment:
This comment is also not covering 100% percent of the cases, right? It definitely correspond to the majority of the cases, but there is a corner case where WFI does not put the core into low-power because the wake-up conditions are satisfied from the beginning. That's in case the function is called with some IRQ pending...

Reply:
Sure, `WFI` will not sleep if there is a pending interrupt; though, the pending interrupt could still be considered a "wake-up interrupt" as it satisfied one of the "wake-up conditions" for the wake-up/not sleeping.

Reply:
The comment could have stated "Clear PRIMASK and flush instruction buffer to immediately service any pending interrupt," but I decided to emphasise/remind that the interrupt being serviced here is (should be) the one that satisfied the wake-up condition.

Reply:
OK